### PR TITLE
Cubeviz: Allow model fitting to operate on spectral Subset

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to allow successful "apply to cube" operation in this workflow:

1. Load cube (left viewer)
2. Generate collapsed cube by sum (middle viewer).
3. Draw X-range region on the spectrum at bottom.
4. Use model fitting plugin, fit on SCI data, use linear model, label it "L1", add the model with default values, type "L1" in equation, label it "LinFit".
5. Select "Subset 1", click "Fit", and it should succeed as expected. See fit shown on spectrum.
6. Click "Apply to cube". Without this patch, it shows error (in notebook but not in lab). With this patch, a fitted cube is shown in right viewer without error (someone familiar with Cubeviz should check if the result is correct or not; I cannot tell).

Fixes the second bug in https://jira.stsci.edu/browse/JDAT-1671

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Comment by pllim** [comment]
I talked to @orifox and here is what the desired behavior _should_ be. I am not sure how this current code works, so probably need @rosteen to advise on the proper fix. I have turned this PR back to draft for now.

1. User select X-range.
2. User defines model.
3. Click "Fit" would first sum all the pixels in the given wavelength, and then fit only the selected wavelength range.
4. Click "Apply to cube" would fit each of the pixels separately (so that's 1 million models for 1000x1000 slice) but only for the selected wavelength range. However, the fitted models would be used to then reconstruct the full cube and display it on the top right.

Currently (without this patch), Step 4 does not work. It crashes here

https://github.com/spacetelescope/jdaviz/blob/d296c6312b020897034e9dd1fc58c84a2559efa5/jdaviz/configs/default/plugins/model_fitting/model_fitting.py#L378

because the X-range subset is not registered as part of `self.app.data_collection`. You can extract it if you dig into `viewer.state.layers`, where `viewer` is the spectrum viewer.

**Comment by rosteen** [comment]
I think what you did here is ok for solving that part of the bugs Ori was seeing - we want `fit to cube` to apply to the original cube, although I wonder if we need to worry about a case where the data isn't the first in the collection. I think the loader code should always put it there though, maybe it's fine.

I do think there are a couple additional things that need to be done to make the model fitting robust for different subset cases:

1. I think that the spectral subsets need to be split out from the `Data` dropdown menu into their own dropdown, as done in the `Collapse` plugin in Cubeviz. This would enable the use case of selecting a spatial region of interest in the image viewer, and then wanting to fit a spectral subset of the collapsed spectrum of that region. 
2. Code would then need to be added to fit only points in the selected spectral region of the selected data, both for the 1D and cube fitting cases. I would think that `specutils.extract_spectral_region` could be used for this. This resolves the current ambiguity in selecting a spectral subset in the `Data` menu (if there are multiple spectra in the spectrum viewer, which are you getting a spectral subset of?).

**Comment by pllim** [comment]
> couple additional things

They sound out of scope. If you think what I have here fixes the traceback that Ori reported, as far as the ticket is concerned, this work is done, and we can open follow-up issue for a more robust fitting. What do you think?

**Comment by rosteen** [comment]
Well, given the name of this PR I'm fine with the changes I laid out being implemented in another PR. But I think they're pretty integral to enabling the workflow Ori was trying, even if not the direct cause of the traceback he posted.

**Comment by pllim** [comment]
Well, I can try but I have no idea how to test whether any additional improvement works because I see no tests for this thing. How do I know the fitted result is actually correct for different workflows?

**Comment by pllim** [comment]
I tried to copy over the Collapse stuff but it doesn't quite work here. Not sure why. Do these errors mean anything to you?

`Error handling response: TypeError: Cannot read property 'takeRecords' of undefined at Function.e.closeListeners`

`widget model not found at x.get_model at u.renderModel`

`Error in event handler: TypeError: Cannot read property 'observe' of undefined`

**Comment by rosteen** [comment]
I didn't encounter that error, I have no insight for you there. But I did see one other error and fixed it; after that this appears to work for me. I'll make a suggestion on the review in a sec.

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py
Comment (rosteen) [inline jdaviz/configs/default/plugins/model_fitting/model_fitting.py:260]:
```suggestion
        self.selected_subset = "None"
        self._window = None
        self.spectral_min = selected_spec.spectral_axis[0].value
        self.spectral_max = selected_spec.spectral_axis[-1].value
        self.spectral_unit = str(selected_spec.spectral_axis.unit)
```

No need to get the whole cube, we can use the collapsed Spectrum1D to get the spectral bounds (this fails as written currently).

**Review by rosteen — CHANGES_REQUESTED** [review]
Mostly looks good to me. Currently waiting for my cube fitting test run to finish, it seems like it takes a good deal longer to run with a spectral subset selected this way - maybe there's a performance hit to using the `window` parameter in `fit_lines`? I'll double check how long it took to run before this change.

**Comment by pllim** [comment]
As per offline discussion, the errors I saw popped up in Lab but not Notebook. 🤷

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py
Comment (pllim) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py:260]: > this fails as written currently

I was gonna apply your suggestion but now I am not sure... LoL.

**Comment by pllim** [comment]
Re: performance -- Not sure what we can do about it. Slow and correct is better than fast and not working. 🤷

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py
Comment (rosteen) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py:260]: Oh sorry, I meant your code fails and this suggestion fixes it.

Comment (pllim) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py]: Ah, gotcha. English can be so confusing to me sometimes! 😸

**Comment by pllim** [comment]
So... I ran these magical incantations and now I think it works in Lab. Not sure what happened... 

```
jupyter labextension install @jupyter-widgets/jupyterlab-manager
jupyter labextension install jupyter-vuetify
```

**Comment by pllim** [comment]
You were not kidding about the performance hit. I had to kill the notebook after I got worried about the wellbeing of my CPU.

**Comment by pllim** [comment]
There is a non-trivial amount of extra processing when `window` is given as `SpectralRegion`:

https://github.com/astropy/specutils/blob/79caf1c604d0fe135d02725a7ef62ef669d1f350/specutils/fitting/fitmodels.py#L430-L462

Is there a way for this plugin to pass in `window` as tuple instead?

**Comment by rosteen** [comment]
I just tested the same thing you just implemented (passing in a tuple window) and it looks like it cut the runtime down from almost 10 minutes to less than a minute. I'm calling that a success!

**Comment by rosteen** [comment]
Unfortunately, the model fitting plugin is shared between all the viz tools...and this is breaking its use in Specviz. It looks like the root of the problem might be that Specviz is using pixel regions that don't have units attached? Error traces below (first from selecting data, second from selecting subset).

```
---------------------------------------------------------------------------
TraitError                                Traceback (most recent call last)
~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/ipyvue/VueTemplateWidget.py in _handle_event(self, _, content, buffers)
     55                 getattr(self, 'vue_' + event)(data, buffers)
     56             else:
---> 57                 getattr(self, 'vue_' + event)(data)
     58 
     59 

~/projects/jdaviz/jdaviz/configs/default/plugins/model_fitting/model_fitting.py in vue_data_selected(self, event)
    254         self.selected_subset = "None"
    255         self._window = None
--> 256         self.spectral_min = selected_spec.spectral_axis[0].value
    257         self.spectral_max = selected_spec.spectral_axis[-1].value
    258         self.spectral_unit = str(selected_spec.spectral_axis.unit)

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/traitlets/traitlets.py in __set__(self, obj, value)
    602             raise TraitError('The "%s" trait is read-only.' % self.name)
    603         else:
--> 604             self.set(obj, value)
    605 
    606     def _validate(self, obj, value):

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/traitlets/traitlets.py in set(self, obj, value)
    576 
    577     def set(self, obj, value):
--> 578         new_value = self._validate(obj, value)
    579         try:
    580             old_value = obj._trait_values[self.name]

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/traitlets/traitlets.py in _validate(self, obj, value)
    608             return value
    609         if hasattr(self, 'validate'):
--> 610             value = self.validate(obj, value)
    611         if obj._cross_validation_lock is False:
    612             value = self._cross_validate(obj, value)

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/traitlets/traitlets.py in validate(self, obj, value)
   2078             value = float(value)
   2079         if not isinstance(value, float):
-> 2080             self.error(obj, value)
   2081         return _validate_bounds(self, obj, value)
   2082 

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/traitlets/traitlets.py in error(self, obj, value, error, info)
    688                     e = "The '%s' trait expected %s, not %s." % (
    689                         self.name, self.info(), describe("the", value))
--> 690                 raise TraitError(e)
    691 
    692     def get_metadata(self, key, default=None):

TraitError: The 'spectral_min' trait of a ModelFitting instance expected a float, not the float32 3799.2686.

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/ipywidgets/widgets/widget.py in _handle_msg(self, msg)
    674                 if 'buffer_paths' in data:
    675                     _put_buffers(state, data['buffer_paths'], msg['buffers'])
--> 676                 self.set_state(state)
    677 
    678         # Handle a state request.

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/ipywidgets/widgets/widget.py in set_state(self, sync_data)
    543                     from_json = self.trait_metadata(name, 'from_json',
    544                                                     self._trait_from_json)
--> 545                     self.set_trait(name, from_json(sync_data[name], self))
    546 
    547     def send(self, content, buffers=None):

~/opt/anaconda3/envs/viz_dev/lib/python3.9/contextlib.py in __exit__(self, type, value, traceback)
    122         if type is None:
    123             try:
--> 124                 next(self.gen)
    125             except StopIteration:
    126                 return False

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/traitlets/traitlets.py in hold_trait_notifications(self)
   1212                 for changes in cache.values():
   1213                     for change in changes:
-> 1214                         self.notify_change(change)
   1215 
   1216     def _notify_trait(self, name, old_value, new_value):

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/ipywidgets/widgets/widget.py in notify_change(self, change)
    604                 # Send new state to front-end
    605                 self.send_state(key=name)
--> 606         super(Widget, self).notify_change(change)
    607 
    608     def __repr__(self):

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/traitlets/traitlets.py in notify_change(self, change)
   1225     def notify_change(self, change):
   1226         """Notify observers of a change event"""
-> 1227         return self._notify_observers(change)
   1228 
   1229     def _notify_observers(self, event):

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/traitlets/traitlets.py in _notify_observers(self, event)
   1262                 c = getattr(self, c.name)
   1263 
-> 1264             c(event)
   1265 
   1266     def _add_notifiers(self, handler, name, type):

~/projects/jdaviz/jdaviz/configs/default/plugins/model_fitting/model_fitting.py in _on_subset_selected(self, event)
    280             spec_sub = self._spectral_subsets[self.selected_subset]
    281             unit = u.Unit(self.spectral_unit)
--> 282             spreg = SpectralRegion.from_center(spec_sub.center.x * unit,
    283                                                spec_sub.width * unit)
    284             self._window = (spreg.lower, spreg.upper)

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/specutils/spectra/spectral_region.py in from_center(cls, center, width)
     80 
     81         if center.unit.physical_type != 'length':
---> 82             return cls(center + width/2, center - width/2)
     83 
     84         return cls(center - width/2, center + width/2)

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/specutils/spectra/spectral_region.py in __init__(self, *args)
     54 
     55         #  Check validity of the input sub regions.
---> 56         if not self._valid():
     57             raise ValueError("SpectralRegion 2-tuple lower extent must be "
     58                              "less than upper extent.")

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/specutils/spectra/spectral_region.py in _valid(self)
    192 
    193         if any(x[0] >= x[1] for x in sub_regions):
--> 194             raise ValueError('Lower bound must be strictly less than the upper bound')
    195 
    196         return True

ValueError: Lower bound must be strictly less than the upper bound
```

**Comment by pllim** [comment]
> TraitError: The 'spectral_min' trait of a ModelFitting instance expected a float, not the float32 3799.2686.

Hmm. Maybe this is why it was set to `Any`. I can change it back. I thought `Float` made more sense.

**Comment by pllim** [comment]
I don't see any error in Specviz with this new commit. Can you please test again? Thanks! 🙏

**Review by rosteen — DISMISSED** [review]
Looks good to me now! Thanks for the work on this, I think this is a huge improvement to the tool.

**Review by javerbukh — APPROVED** [review]
Followed the steps to reproduce the bug and the app worked as expected. Looks good!

**Comment by rosteen** [comment]
Re-requested a review from myself so we don't merge seeing two approvals, since a bug was brought up offline (can't select 2D subsets in the `Data` menu as expected).

**Comment by pllim** [comment]
@rosteen , re-requesting self review does not re-block merge. I have _dismissed_ your previous approval. Merge is now blocked again. FYI.

Also, I still don't grok this plugin, so I post the dropdown menu behaviors below and you can tell me if they are correct.

Here is the Data dropdown for the _viewer_ after playing around with the plugin:

Screenshot 2021-08-24 160715

Here is the Data dropdown for the _model fitting plugin_ at the same state:

Screenshot 2021-08-24 160803

Here is the Subset dropdown for the _model fitting plugin_ at the same state:

Screenshot 2021-08-24 160819

Bonus: I don't even know what this selection below means but it is allowed, though even you click "apply to cube," nothing happens because in the actual code, it is a no-op if the Data selected in plugin is not a cube.

Screenshot 2021-08-24 160859

**Comment by pllim** [comment]
I am not sure how, but I think this now fixes [the first two bugs Ori reported in JIRA](https://jira.stsci.edu/browse/JDAT-1671?focusedCommentId=544823&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-544823). Please double check.

**Comment by rosteen** [comment]
I was going to ask you to rebase so we see that test pass...but then the failing dev test will pop up. I think I'm just going to approve.

**Review by rosteen — APPROVED** [review]

**Comment by rosteen** [comment]
Actually, the dev test is an allowed failure right? If you rebase @pllim we can probably unblock this and merge it.

**Comment by pllim** [comment]
Yes, dev job failing does not block merge and the current failure is unrelated. I rebased. Thanks for the approval!

**Comment by pllim** [comment]
And... 8 points down, yay lol

**Pull request merged**
